### PR TITLE
Added code for Readers-Writers problem in OS using Semaphores

### LINKED ISSUE #1874: Add code for readers-writer problem using semaphore.
I would like to add the cpp implementation of this code. Please assign this.
@raghav-dalmia @MastersAbh @asha15 @singh-shreya6 @SakshiShreya

### PR DESCRIPTION
**Pull request by Xxoliaxx**
Fixes #1874

#### Checklist:
- [x] 4 space indentation
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

#### Changes proposed in this pull request:
- Added the implementation of Readers Writers problem in OS using semaphores in C++

#### Languages Used:
- C++

#### Files added:
- Readers_Writers_Problem.cpp
- Readme.md

### CODE REVIEW

**Comment by SakshiShreya** [comment]
Please mention the issue number in the first line of the PR template after #.

**Comment by Xxoliaxx** [comment]
I've made the necessary changes. Please let me know if there are any other changes required. @SakshiShreya

File: Readers_Writers_Problem/Readers_Writers_Problem.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	sem_wait(&w);		//Waits for write block to be released
+	d++;
+	cout<<"Writer "<<i<<" writes "<<d<<endl;
+	sleep(1);
+	cout<<"Done";
+	sem_post(&w);		//Release the write block
+}
+
+int main(){
+	long i,b;
```
Comment (jainaman224) [inline Readers_Writers_Problem/Readers_Writers_Problem.cpp]: Incorrect indentation. Use 4 space indentation.

Comment (jainaman224) [reply Readers_Writers_Problem/Readers_Writers_Problem.cpp]: Fix this throughout the file.

**Comment by jainaman224** [comment]
@Xxoliaxx Are you working on it?

**Comment by Xxoliaxx** [comment]
@jainaman224 Yes I am working on it.

**Comment by Xxoliaxx** [comment]
@jainaman224 I have made the necessary changes. Please review.

File: Readers_Writers_Problem/Readers_Writers_Problem.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
+//Code for Readers-Writers problem using semaphores and threads
+#include<bits/stdc++.h>
+#include<semaphore.h>
+#include<unistd.h>
+#include<pthread.h>
+using namespace std;
+
+sem_t m,w;			//m is mutex and w is write_block semaphore
```
Comment (somya-kapoor) [inline Readers_Writers_Problem/Readers_Writers_Problem.cpp]: Add a space after comma in the whole file

File: Readers_Writers_Problem/Readers_Writers_Problem.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+int d = 0, readers = 0;         //d is a variable that stores what has to be read or written and readers is for number of readers
```
Comment (somya-kapoor) [inline Readers_Writers_Problem/Readers_Writers_Problem.cpp]: Move comments above the statements in the whole file.

File: Readers_Writers_Problem/Readers_Writers_Problem.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    sem_wait(&w);		//Waits for write block to be released
+    d++;
+    cout << "Writer " << i << " writes " << d << endl;
+    sleep(1);
+    sem_post(&w);		//Release the write block
+}
+
+int main()
+{
+    long i,b;
```
Comment (somya-kapoor) [inline Readers_Writers_Problem/Readers_Writers_Problem.cpp]: Add space after comma

File: Readers_Writers_Problem/Readers_Writers_Problem.cpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    pthread_t rthreads[5],wthreads[5];
+    sem_init(&m,0,1);	//Initializes mutex
+    sem_init(&w,0,1);	//Initializes write block semaphore
+  	
+    //Loop for creating threads
+    for(i = 0;i <= 2;i++)
```
Comment (somya-kapoor) [inline Readers_Writers_Problem/Readers_Writers_Problem.cpp]: Add space after ;

**Comment by Xxoliaxx** [comment]
@somya-kapoor I have made the changes you've asked for. Do review.

File: Readers_Writers_Problem/Readers_Writers_Problem.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        pthread_create(&rthreads[i], NULL, reader, (void *)i);
+    }
+    //Loop for joining threads
+    for(i = 0; i <= 2; i++)
+    {
+        pthread_join(wthreads[i], NULL);
+        pthread_join(rthreads[i], NULL);
+    }
+    return 0;
+}
```
Comment (somya-kapoor) [inline Readers_Writers_Problem/Readers_Writers_Problem.cpp]: Add sample input and output at the end of the file, don't need to create a separate file for it

**Comment by Xxoliaxx** [comment]
@somya-kapoor Done

File: Readers_Writers_Problem/Readers_Writers_Problem.cpp [hunk 7]
```diff
@@ -0,0 +1,85 @@
+//Code for Readers-Writers problem using semaphores and threads
```
Comment (somya-kapoor) [inline Readers_Writers_Problem/Readers_Writers_Problem.cpp]: Add space after //.
Do it in the whole file

**Comment by somya-kapoor** [comment]
@Xxoliaxx After making the changes, please squash the commits

**Comment by Xxoliaxx** [comment]
@somya-kapoor I have made the changes but I am not able to find a way to squash the commits. Can you help me?

**Comment by somya-kapoor** [comment]
@Xxoliaxx For reference http://blog.steveklabnik.com/posts/2012-11-08-how-to-squash-commits-in-a-github-pull-request

**Comment by somya-kapoor** [comment]
@Xxoliaxx Are you working on it?

**Comment by Xxoliaxx** [comment]
@somya-kapoor I have made the changes and I texted you on slack about my problem with squashing commits a day ago already.

**Comment by somya-kapoor** [comment]
@Xxoliaxx Are you working on it?

**Comment by Xxoliaxx** [comment]
Please check slack @somya-kapoor

**Comment by somya-kapoor** [comment]
I have replied to you.
@Xxoliaxx Are you working on it?

**Comment by somya-kapoor** [comment]
@Xxoliaxx Are you working on it?

**Comment by Xxoliaxx** [comment]
I tried everything, it's not getting squashed. Otherwise the code and everything is done.

**Pull request closed without merging**